### PR TITLE
Add COWC trainer

### PR DESCRIPTION
**Pull request by adamjstewart**
- [x] Fix bug in train/test list file encoding
- [x] Add COWC Counting trainer
- [ ] Add COWC Detection trainer? (might wait until refactoring is complete)
- [x] Add tests
- [x] Find out what the state-of-the-art performance should be
- [x] Achieve state-of-the-art performance (or come close)

### CODE REVIEW

File: torchgeo/datasets/cowc.py [hunk 1]
```diff
@@ -101,7 +101,9 @@ def __init__(
         self.images = []
         self.targets = []
         with open(
-            os.path.join(self.root, self.filename.format(split)), newline=""
+            os.path.join(self.root, self.filename.format(split)),
+            encoding="utf-8-sig",
```
Comment (adamjstewart) [inline torchgeo/datasets/cowc.py:105]: The files we test on were ASCII, but the actual downloaded files were UTF-8 Unicode (with BOM) text. This would result in some extra header garbage being added to the name of the first image filepath. I updated the test files to be UTF-8 Unicode (with BOM) text so we can properly test this.

**Comment by adamjstewart** [comment]
Table 3 of the original 2016 paper (https://doi.org/10.1007/978-3-319-46487-9_48) shows an RMSE of 0.657 with ResCeption.

**Comment by adamjstewart** [comment]
Haven't tuned hyperparams yet, but the default hyperparams get us a test RMSE of 0.706 with just a ResNet-18.

File: torchgeo/trainers/cowc.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        batch_size: int = 64,
+        num_workers: int = 4,
+        **kwargs: Any,
+    ) -> None:
+        """Initialize a LightningDataModule for COWC Counting based DataLoaders.
+
+        Args:
+            root_dir: The ``root`` arugment to pass to the
+                COWCCounting Datasets classes
+            seed: The seed value to use when doing the sklearn based GroupShuffleSplit
```
Comment (calebrob6) [inline torchgeo/trainers/cowc.py:34]: seed is not used here

File: torchgeo/trainers/cowc.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+            batch_size: The batch size to use in all created DataLoaders
+            num_workers: The number of workers to use in all created DataLoaders
+            api_key: The RadiantEarth MLHub API key to use if the dataset needs to be
```
Comment (calebrob6) [inline torchgeo/trainers/cowc.py:37]: Neither is api_key

File: torchgeo/trainers/cowc.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        """Transform a single sample from the Dataset.
+
+        Args:
+            sample: dictionary containing image and target
+
+        Returns:
+            preprocessed sample
+        """
+        sample["image"] = sample["image"] / 255.0  # scale to [0,1]
+        sample["label"] = sample["label"].float()
```
Comment (calebrob6) [inline torchgeo/trainers/cowc.py:56]: Is this key correct for regression datasets?

File: torchgeo/trainers/cowc.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+        Args:
+            stage: stage to set up
+        """
+        train_val_dataset = COWCCounting(
+            self.root_dir, split="train", transforms=self.custom_transform
+        )
+        self.test_dataset = COWCCounting(
+            self.root_dir, split="test", transforms=self.custom_transform
+        )
+        self.train_dataset, self.val_dataset = random_split(
```
Comment (calebrob6) [inline torchgeo/trainers/cowc.py:82]: could a seed be passed here to make train/val splits reproducible?

File: conf/task_defaults/cowc_counting.yaml [hunk 6]
```diff
@@ -0,0 +1,9 @@
+experiment:
+  task: cowc_counting
+  module:
+    model: resnet18
+    learning_rate: 1e-3
```
Comment (calebrob6) [inline conf/task_defaults/cowc_counting.yaml:5]: what happens if you start with imagenet weights?

**Review by calebrob6 — CHANGES_REQUESTED** [review]

**Comment by calebrob6** [comment]
What do the tensorboard train/val loss curves look like for this?

File: torchgeo/datasets/cowc.py (hunk 1 above)
Comment (calebrob6) [reply torchgeo/datasets/cowc.py:105]: good catch!

File: torchgeo/trainers/cowc.py (hunk 4 above)
Comment (adamjstewart) [reply torchgeo/trainers/cowc.py:56]: Yes, all scalar values use "label" and all matrices use "mask".

Comment (adamjstewart) [reply torchgeo/trainers/cowc.py:56]: See #213 for more discussion

File: conf/task_defaults/cowc_counting.yaml (hunk 6 above)
Comment (adamjstewart) [reply conf/task_defaults/cowc_counting.yaml:5]: Testing now

**Comment by adamjstewart** [comment]
> What do the tensorboard train/val loss curves look like for this?

How do I view this? Can it only be viewed while training or can it be viewed from the final checkpoint? I've never used tensorboard before...

**Comment by isaaccorley** [comment]
I don't believe we are saving any metrics in the checkpoints so you would need to check the tensorboard logs. They can be viewed after training since they are just log files. Just run tensorboard --logdir=path/to/logs/

File: torchgeo/trainers/cowc.py (hunk 4 above)
Comment (calebrob6) [reply torchgeo/trainers/cowc.py:54]: Ah nice

**Comment by adamjstewart** [comment]
Okay, this finally finished. With ImageNet pretraining, we get an RMSE of 0.631 and an MAE of 0.316.

**Review by calebrob6 — APPROVED** [review]

**Comment by calebrob6** [comment]
Nicely done! And congrats on your first trainer!

**Comment by adamjstewart** [comment]
Second trainer 😛
